acceptance/tests/resource/user/should_manage_groups.rb

```diff
@@ -0,0 +1,124 @@
+
+  # TODO: This should be added to Beaker
+  def assert_matching_arrays(actual, expected, message = "")
+    assert_equal(actual.sort, expected.sort, message)
+  end
+
+  agents.each do |agent|
+    groups = 5.times.collect { random_name }
+    groups.each { |group| agent.group_absent(group) }
+
+    # We want to ensure that Beaker destroys the user first before
+    # the groups. Otherwise the teardown step will fail b/c we will
+    # be trying to remove the user's primary group before removing
+    # the user.
+    user = random_name
+    agent.user_absent(user)
+    teardown { agent.user_absent(user) }
+
+    step 'Creating the Groups' do
+      groups.each do |group|
+        agent.group_present(group)
+        teardown { agent.group_absent(group) }
+      end
+    end
+
+    user_groups = [groups[0], groups[1]]
+    primary_group = groups[2]
+
+    step 'Ensure that the user is created with the specified groups' do
+      # We use inclusive membership to ensure that the user's only a member
+      # of our groups and no other group.
+      manifest = user_manifest(user, groups: user_groups, gid: primary_group, membership: :inclusive)
+      apply_manifest_on(agent, manifest)
+      assert_matching_arrays(groups_of(agent, user), user_groups, "The user was not successfully created with the specified groups!")
+    end
+
+    step "Verify that Puppet errors when one of the groups does not exist" do
+      manifest = user_manifest(user, groups: ['nonexistent_group'])
+      apply_manifest_on(agent, manifest) do |result|
+        assert_match(/Error:.*#{user}/, result.stderr, "Puppet fails to report an error when one of the groups in the groups property does not exist")
+      end
+    end
+
+    primary_group = groups[3]
+    step "Verify that modifying the primary group does not change the user's groups" do
+      manifest = user_manifest(user, gid: primary_group)
+      apply_manifest_on(agent, manifest)
+      assert_matching_arrays(groups_of(agent, user), user_groups, "The user's groups changed after modifying the primary group")
+    end
+
+    step "Verify that Puppet noops when the user's groups are already set" do
+      manifest = user_manifest(user, groups: user_groups)
+      apply_manifest_on(agent, manifest, catch_changes: true)
+      assert_matching_arrays(groups_of(agent, user), user_groups, "The user's groups somehow changed despite Puppet reporting a noop")
+    end
+
+    step "Verify that Puppet enforces minimum group membership" do
+      new_groups = [groups[2], groups[4]]
+
+      manifest = user_manifest(user, groups: new_groups, membership: :minimum)
+      apply_manifest_on(agent, manifest)
+
+      user_groups += new_groups
+      assert_matching_arrays(groups_of(agent, user), user_groups, "Puppet fails to enforce minimum group membership")
+    end
+
+    if agent['platform'] =~ /osx/
+      skip_test "User provider on OSX fails to enforce inclusive group membership, so we will skip that test until this is fixed."
```

Comment:
This is a legitimate bug that I found. I tried to see if it was b/c reading the `/etc/group` file does not work on OSX platforms, but I found that both `dscacheutil -q group` and `puppet resource user <user>` return the same list of groups for the user. Is there a ticket filed for this? If not, I can file one.

Reply:
I will file a ticket for this and then reference it in the test.